src/sshfs_mount/sshfs_mount.cpp

```diff
@@ -27,36 +27,78 @@
+const std::string snap_path_key{"SNAP="};
+
+auto run_cmd(mp::SSHSession& session, std::string&& cmd)
 {
     auto ssh_process = session.exec(cmd);
     if (ssh_process.exit_code() != 0)
-        error_handler(ssh_process);
-    return ssh_process.read_std_output();
-}
+        throw std::runtime_error(ssh_process.read_std_error());
 
-auto run_cmd(mp::SSHSession& session, std::string&& cmd)
-{
-    auto error_handler = [](mp::SSHProcess& proc) { throw std::runtime_error(proc.read_std_error()); };
-    return run_cmd(session, std::forward<std::string>(cmd), error_handler);
+    return ssh_process.read_std_output() + ssh_process.read_std_error();
 }
 
-void check_sshfs_exists(mp::SSHSession& session)
+auto get_sshfs_exec_and_options(mp::SSHSession& session)
 {
-    auto error_handler = [](mp::SSHProcess& proc) {
-        mpl::log(mpl::Level::warning, category,
-                 fmt::format("Unable to determine if 'sshfs' is installed: {}", proc.read_std_error()));
-        throw mp::SSHFSMissingError();
-    };
-
-    run_cmd(session, "which sshfs", error_handler);
+    std::string sshfs_exec;
+
+    try
+    {
+        // Prefer to use Snap package version first
+        std::istringstream sshfs_env{run_cmd(session, "sudo multipass-sshfs.env")};
+
+        auto ld_library_path = mp::utils::match_line_for(sshfs_env, ld_library_path_key);
+        auto snap_path = mp::utils::match_line_for(sshfs_env, snap_path_key);
+        snap_path = snap_path.substr(snap_path_key.length());
+
+        sshfs_exec = fmt::format("env {} {}/bin/sshfs", ld_library_path, snap_path);
+    }
+    catch (const std::exception& e)
+    {
+        mpl::log(mpl::Level::debug, category, fmt::format("'multipass-sshfs' snap package is not installed: {}", e.what()));
+
+        // Fallback to looking for distro version if snap is not found
+        try
+        {
+            sshfs_exec = run_cmd(session, "sudo which sshfs");
+        }
+        catch (const std::exception& e)
+        {
+            mpl::log(mpl::Level::warning, category,
+                     fmt::format("Unable to determine if 'sshfs' is installed: {}", e.what()));
+            throw mp::SSHFSMissingError();
+        }
+    }
+
+    sshfs_exec = mp::utils::trim_end(sshfs_exec);
+
+    std::istringstream version_info{run_cmd(session, fmt::format("sudo {} -V", sshfs_exec))};
+
+    sshfs_exec += " -o slave -o transform_symlinks -o allow_other";
+
+    auto fuse_version_line = mp::utils::match_line_for(version_info, fuse_version_string);
+    if (!fuse_version_line.empty())
+    {
+        // split on the fuse_version_string along with 0 or more colons
+        auto tokens = mp::utils::split(fuse_version_line, fmt::format("{}:* ", fuse_version_string));
+        auto fuse_version = tokens[1];
+
+        // The option was removed in libfuse 3.0
+        if (!fuse_version.empty() && (version::Semver200_version(fuse_version) < version::Semver200_version("3.0.0")))
+            sshfs_exec += " -o nonempty";
```

Comment:
Error that we failed to parse the version?

Reply:
Yes, and it's really necessary for the testing.

Reply:
Oh, you mean throw an error and/or log an error?

Reply:
Yes :)

Reply:
Hmm, if I throw here, I'll have to modify the tests quite a bit to mock returning some valid version number here.  I can certainly add the logging, but maybe we just leave in the option as default if we can't figure out the version number.  WDYT?

Reply:
Oh, I think `semver` will throw if we try to use an invalid version number.

Reply:
Yeah, just logging the error is fine, but: keep the option _in_, our _out_? I think I'm leaning _out_, since it's more compatible, and it's not very dangerous if it's left out?

Reply:
If we don't specify `-o nonempty` on a directory that isn't empty, then `sshfs` will error out.  Not dangerous by any means.

Reply:
Ok, I modified the logic a bit and logging.  I think the option should only be added if we can be sure the version is < 3.0.0, otherwise leave it out.